lib/wasi/src/lib.rs

```diff
@@ -72,31 +74,62 @@ pub enum WasiError {
 }
 
 /// The environment provided to the WASI imports.
-#[derive(Debug, Clone, WasmerEnv)]
-pub struct WasiEnv {
-    /// Shared state of the WASI system. Manages all the data that the
-    /// executing WASI program can see.
-    ///
-    /// Be careful when using this in host functions that call into Wasm:
-    /// if the lock is held and the Wasm calls into a host function that tries
-    /// to lock this mutex, the program will deadlock.
-    pub state: Arc<Mutex<WasiState>>,
+#[derive(Derivative)]
+#[derivative(Debug)]
+#[derive(Clone, WasmerEnv)]
+pub struct WasiThread {
+    /// ID of this thread
+    id: u32,
+    /// Provides access to the WASI environment
+    env: WasiEnv,
     #[wasmer(export)]
     memory: LazyInit<Memory>,
 }
 
-impl WasiEnv {
-    pub fn new(state: WasiState) -> Self {
-        Self {
-            state: Arc::new(Mutex::new(state)),
-            memory: LazyInit::new(),
+/// The wasi thread dereferences into the wasi environment
+impl Deref for WasiThread {
+    type Target = WasiEnv;
+
+    fn deref(&self) -> &WasiEnv {
+        &self.env
+    }
+}
+
+impl WasiThread {
+    // Returns the unique ID of this thread
+    pub fn thread_id(&self) -> u32 {
+        self.id
+    }
+
+    // Yields execution
+    pub fn yield_now(&self) {
+        if let Some(callback) = self.on_yield.as_ref() {
+            callback(self);
+        }
+        std::thread::yield_now();
+    }
+
+    // Sleeps for a period of time
+    pub fn sleep(&self, duration: std::time::Duration) {
+        let start = std::time::Instant::now();
+        loop {
+            self.yield_now();
+            let delta = std::time::Instant::now() - start;
+            if delta > duration {
+                break;
+            }
+            let remaining = duration - delta;
+            std::thread::sleep(remaining.min(std::time::Duration::from_millis(10)));
+            if std::time::Instant::now() - start >= duration {
+                break;
+            }
```

Comment:
`Duration` math ops might panic. Better use the `checked_` versions which return `None` on error/bug:

https://doc.rust-lang.org/std/time/struct.Instant.html#method.checked_duration_since